### ISSUE TITLE
Some menus are blank

### ISSUE THREAD

**Issue by IdleSol** [CleverRaven/Cataclysm-DDA#71994]
### Describe the bug

After update #65709. @katemonster33

Two versions were compared:
- [2024-02-26-0558](https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2024-02-26-0558)
- [2024-02-26-0341](https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2024-02-26-0341)

See the screenshots for details

### Attach save file

N/A

### Steps to reproduce

N/A

### Expected behavior

N/A

### Screenshots

![11](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/af839826-7635-4624-aa6f-130cfeb7b0f2)
![12](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/4b08815f-b07c-4d09-8e80-afa27a4f0f5c)
![13](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/bb3e29bf-4592-4f75-88a1-56b27c4d4312)
![14](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/758a64ee-da07-4785-a0aa-f8293136e406)
![15](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/b98228c8-8abf-45c4-808a-8836f6103b9c)
![16](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/8a4a3478-40a5-481c-a9d1-efb25bcc9995)

### Versions and configuration

cdda-windows-tiles-x64-2024-02-26-0558

### Additional context

If I find anything else, I'll add to the posts.

**Comment by IdleSol** [comment]
When trying to select an AR-15 

![1](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/68db0200-d44c-4afd-9b12-60f79dcab36e)

The game is running through Portproton (ArchLinux)

- OS: Windows
    - OS Version: 10.0.18362 (21H1)
- Game Version: 6e1e928 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

**Comment by NetSysFire** [comment]
I think your very first screenshot shows an existing potential "inconsistency" that may be unaffected by the recent imgui addition where the version is placed after any ascii art that may be displayed

**Comment by katemonster33** [comment]
I've confirmed most of these screenshots, What view are you in when you try to "select an AR-15"?

**Comment by NetSysFire** [comment]
>Or I was unlucky.

It is rng-based. That ascii art only shows up sometimes.

**Comment by IdleSol** [comment]
@katemonster33 
In the game itself, I haven't seen any problems with the interface. But I was only looking at the ```Shift+1..0, -, =``` windows.

**Comment by katemonster33** [comment]
I've narrowed it down to this commit causing the issue https://github.com/CleverRaven/Cataclysm-DDA/commit/f37380ed3312d82d3c61533ccfcffbc38097f5b4 maybe @db48x has some wisdom? hate to just revert the commit

**Comment by IdleSol** [comment]
![1](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/69fa84ff-3210-494e-b8a1-f8674fdefe6a)

This is not your mistake, it was like this in previous versions. But maybe you can fix it:
![test](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/a079d577-c27d-4466-bbaf-c2397f76470e)

**Comment by IdleSol** [comment]
![1](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/da869083-61fa-480a-993f-3a140cc377ec)

As well as other items from the "main menu" (press esc in the game):
- ```4``` Auto pickup manager
- ```5``` Auto notes manager 
- ```6``` Safe mode manager
- ```7``` Distractions manager (see screenshot)
- ```8``` Color manager

I didn't take screenshots for all of these points. They all have a header problem.

**Comment by katemonster33** [comment]
> When trying to select an AR-15
> 
> ![1](https://private-user-images.githubusercontent.com/155764584/307832253-68db0200-d44c-4afd-9b12-60f79dcab36e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5ODU4NzMsIm5iZiI6MTcwODk4NTU3MywicGF0aCI6Ii8xNTU3NjQ1ODQvMzA3ODMyMjUzLTY4ZGIwMjAwLWQ0NGMtNGFmZC05YjEyLTYwZjc5ZGNhYjM2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyNlQyMjEyNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZWIzNmU1MTg0MWM1NWZiNmUzY2QxMzJlMWRlOWFiNzEyZTkxYjY3NjY1MWQzZjY4NzRhMDZmNDkzZmQxODkzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.0eRJriZsqLHIXFC0bkOlcjckDDuv_t6LRNwh4qsvk2Y)
> 
> The game is running through Portproton (ArchLinux)
So running on Wine? This didn't happen for me on Windows. Can you try on a native Linux build?

**Comment by db48x** [comment]
I can reproduce it. Curiously enough, we are drawing spaces on those lines that are missing.

**Comment by db48x** [comment]
I haven't figured it out, but I have narrowed it down a bit. In `user_interface::show` in the file `auto_pickup.cpp` there is an `on_redraw` callback. This callback draws the autopickup manager window in three parts. The border is drawn into `w_border`, the header into `w_header`, and the rules in`w`. If I comment out everything after the call to `wnoutrefresh( w_header );`, then the header is correctly drawn. In fact, after drawing and refreshing `w_header`, it goes back and adds a scrollbar to `w_border`. If I comment out just the second call to `wnoutrefresh( w_border );` then the header is also still visible. So it seems that refreshing this window is erasing the header.

And moving the call to `draw_scrollbar` up to before the header is ever drawn also fixes it. I suspect the other windows all have the same problem. We could fix them all by rearranging them to add the scrollbar earlier, or we could delve deeper into `cursesport` to see why it is erasing the header window (which is overlapping with the border window). I need to get some sleep, so I leave it to you @katemonster33. I’ll check back in tomorrow evening.

**Comment by katemonster33** [comment]
> I can reproduce it. Curiously enough, we are drawing spaces on those lines that are missing.

Can you print the stack trace?

**Comment by db48x** [comment]
I guess it’s just because every window has its own `line` vector. It calls `wnoutrefresh( w_header );` which draws `w_header->line`, then it calls `wnoutrefresh( w_border );` which draws `w_border->line` over top of it.

**Comment by db48x** [comment]
Another fix would be to give each scrollbar their own window that is only one column wide. Then it can be drawn any time you want without clearing the whole window. But that’s a more invasive change than just rearranging the order that windows are drawn in.

**Comment by katemonster33** [comment]
I'm coming up with nothing. We may need to simply put the framebuffer back in for now and come up with some solution for letting it work with imgui... maybe invalidate the framebuffer on every redraw when imgui is on screen. 

was hoping we could just remove that commit temporarily and make a new draft PR with the commit to fix master for now.... waiting on an answer to see if that is possible.

**Comment by IdleSol** [comment]
>So running on Wine? This didn't happen for me on Windows. Can you try on a native Linux build?

Unfortunately, I can't. The available version is from January 9 (0.G.2024.01.09-1). And for building from source, I don't have enough knowledge.

**Comment by NetSysFire** [comment]
There are available builds for Linux built every day, e.g cdda-linux-tiles. https://github.com/CleverRaven/Cataclysm-DDA/releases

**Comment by IdleSol** [comment]
Here is the result on another windows 10 computer.

- OS: Windows
    - OS Version: 10.0.19045.4046 (22H2)
- Game Version: 6e1e928 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]
![1](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/81121fcb-7338-4f49-816c-a37057abe092)
Same as on the virtual machine

If you press ```alt+f4``` while the game is loading:
![2](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/bb9b88f7-7e6f-4568-87fe-336d49a5fc64)
Once the window is closed, a second window will appear:
![3](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/cdf51389-3bcf-4fa8-8cab-3eacb7c86258)

I've been paying attention. In my settings: ```Use ImGui IU: false```. Doesn't that mean everything stays as it was? Why did I have all these errors with windows then?

Enabling the UI, does not change the window header errors.

**Comment by IdleSol** [comment]
Additional information.

If the game is run through PortProton, the error (with imGui) causes the game to crash.
If the game is run on Windows, the error (with imGui) does not cause a game crash (if you press skip).

This error (with imGui) occurs before the game error message. For example: #71451

Calling the test error message, does not cause an error with imGui.

If you get a game error and click ignore error messages, then repeating the error, does not cause an error with imGui.

Example (win10):
1. Open the debug menu
2. Spawn an item ```s```/```w```
3. Enter "ar-15" in the search bar
4.1. There is an error with imGui: line 9113
3.2. Click skip
3.3. Get an error: "loaded a nun or ammo_data() is nullptr"
3.4. Press ```I```
3.5. Сlose the spawn window
5. Repeat steps 2 and 3. No errors

The problem is that not every window causes an error to appear with imGui.

For example, disconnecting a modification from the m4 carbine causes a "loaded a nun or ammo_data() is nullptr" error, but does not cause an imGui error.

**Comment by db48x** [comment]
Ok, it is quite evident that there is more than one bug here. We need to split this up into multiple bug reports.

- [ ] Separately report ammo_data/loaded nun errors (which is probably not ImGui related)
- [ ] Separately report SDL version mismatch (now requires SDL >=2.0.17)
- [ ] Separately report ImGui uninitialized errors on Windows (it is vital to include the crash log here, for the stack trace)

We’ll leave this bug report for the windows that are missing some lines, usually headers, but also sometimes descriptions.

**Comment by IdleSol** [comment]
> Separately report ammo_data/loaded nun errors (which is probably not ImGui related)

- #71781

**Comment by db48x** [comment]
And #72022

**Comment by db48x** [comment]
At the moment I am inclined to simply look for all places where `draw_scrollbar` is called, and move that call up so that it happens just after the window border is drawn. It’s a little bit like rearranging the deck chairs on the Titanic, since all of that code will be rewritten in ImGui and ImGui will handle the scrollbars for us, but it’s an easy fix. I’ll do it myself this evening, unless someone else does something about it before then.

**Comment by db48x** [comment]
> … Maybe instead of clearing the window area before drawing a window, only erase locations in the window that have been printed to? …

I was thinking about this all day and eventually realized that we could just clear the window line by line, skipping the lines that aren’t `touched`, instead of clearing the whole window in one go.

PR incoming.

**Comment by inverimus** [comment]
I now get a blank screen when starting the game that then draws the main menu on any keypress. Is that related to this?

**Comment by db48x** [comment]
Can you give us some more details? What OS? Is the ImGui interface turned on? Can you share a screenshot?

**Comment by db48x** [comment]
On second thought, a screenshot of a blank screen might not tell us much :) A video might explain it better. Is the main menu displayed normally once you press a key, or does it go away again, or what?

**Comment by katemonster33** [comment]
> I now get a blank screen when starting the game that then draws the main menu on any keypress. Is that related to this?

Possibly, not likely. Open a new issue with all your platform and configuration details please

**Comment by IdleSol** [comment]
>Separately report ImGui uninitialized errors on Windows (it is vital to include the crash log here, for the stack trace)

#72047
I think it's the same bug. Only in Linux.

**Comment by IdleSol** [comment]
@db48x @katemonster33 

>Separately report ImGui uninitialized errors on Windows (it is vital to include the crash log here, for the stack trace)

cdda-windows-tiles-x64-2024-03-01-0629. Fixed game crash.
